lib/shelley/src/Cardano/Wallet/Shelley/MinimumUTxO.hs

```diff
@@ -0,0 +1,188 @@
+--   inclusion in a transaction output.
+--
+-- This function returns a value that is specific to a given Shelley-based era.
+-- Importantly, a value that is valid in one era will not necessarily be valid
+-- in another era.
+--
+computeMinimumCoinForShelleyBasedEra
+    :: HasCallStack
+    => MinimumUTxOForShelleyBasedEra
+    -> TokenMap
+    -> Coin
+computeMinimumCoinForShelleyBasedEra
+    (MinimumUTxOForShelleyBasedEra era pp) tokenMap =
+        extractResult $
+        Cardano.calculateMinimumUTxO era
+            (embedTokenMapWithinPaddedTxOut era tokenMap)
+            (Cardano.fromLedgerPParams era pp)
+  where
+    extractResult :: Either Cardano.MinimumUTxOError Cardano.Value -> Coin
+    extractResult = \case
+        Right value ->
+            -- We assume that the returned value is a non-negative ada quantity
+            -- with no other assets. If this assumption is violated, we have no
+            -- way to continue, and must raise an error:
+            value
+                & unsafeValueToLovelace
+                & unsafeLovelaceToWalletCoin
+        Left e ->
+            -- The 'Cardano.calculateMinimumUTxO' function should only return
+            -- an error if a required protocol parameter is missing.
+            --
+            -- However, given that values of 'MinimumUTxOForShelleyBasedEra'
+            -- can only be constructed by supplying an era-specific protocol
+            -- parameters record, it should be impossible to trigger this
+            -- condition.
+            --
+            -- Any violation of this assumption indicates a programming error.
+            -- If this condition is triggered, we have no way to continue, and
+            -- must raise an error:
+            --
+            error $ unwords
+                [ "computeMinimumCoinForUTxO:"
+                , "unexpected error:"
+                , show e
+                ]
+
+-- | Embeds a 'TokenMap' within a padded 'Cardano.TxOut' value.
+--
+-- When computing the minimum UTxO quantity for a given 'TokenMap', we do not
+-- have access to an address or to an ada quantity.
+--
+-- However, in order to compute a minimum UTxO quantity through the Cardano
+-- API, we must supply a 'TxOut' value with a valid address and ada quantity.
+--
+-- It's imperative that we do not underestimate minimum UTxO quantities, as
+-- this may result in the creation of transactions that are unacceptable to
+-- the ledger. In the case of change generation, this would be particularly
+-- problematic, as change outputs are generated automatically, and users do
+-- not have direct control over the ada quantities generated.
+--
+-- However, while we cannot underestimate minimum UTxO quantities, we are at
+-- liberty to moderately overestimate them.
+--
+-- Since the minimum UTxO quantity function is monotonically increasing in the
+-- serialized length of its input, if we supply a 'TxOut' with an address and
```

Comment:
> monotonically increasing in the serialized length of its input,

Are you 100% sure about this? I.e. are you 100% sure there are no exotic pairs of pre-babbage outputs such that
```
serializedLength out1 < serializedLength out2 && minUTxOValue out1 > minUTxOValue out2
```

Perhaps replacing "serialized length of its input" with "serialised length of the ada quantity and address" might be what we really mean.

Reply:
Ah that's a good point, thanks for raising it!

I'll adjust this comment as you suggest.

Reply:
@Anviking Fixed in eb446c64aa6febf3fd3f06f950ef904f2ec019c4.